tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.ElevatedServer/Timer.h

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation and Contributors
+// Licensed under the MIT license.
+
+#include "pch.h"
+
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <future>
+#include <mutex>
+#include <optional>
+
+using CallbackFunction = void (*)();
+
+class Timer
+{
+public:
+    // Cleanup functions
+    static void Discard(std::unique_ptr<Timer> timer);
+    static void WaitForAllDiscardedTimersToDestruct();
+
+    Timer(std::chrono::seconds seconds, CallbackFunction callback)
+    {
+        m_startTime = std::chrono::steady_clock::now();
+        m_duration = seconds;
+        m_callback = std::move(callback);
+        m_timerThreadFuture = std::async(std::launch::async, &Timer::TimerThread, this);
+    }
+
+    Timer(Timer&& other) noexcept = default;
+    Timer& operator=(Timer&& other) noexcept = default;
+
+    Timer(const Timer&) = delete;
+    Timer& operator=(const Timer&) = delete;
+
+    void Cancel()
+    {
+        auto lock = std::scoped_lock(m_mutex);
+        m_cancelled = true;
+    }
+
+    int64_t TimeLeftInSeconds()
+    {
+        auto lock = std::scoped_lock(m_mutex);
+        if (m_cancelled)
+        {
+            return 0;
+        }
+        auto now = std::chrono::steady_clock::now();
+        auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - m_startTime);
+
+        auto left = m_duration.count() - elapsed.count();
+        return std::max(left, 0ll);
+    }
+
+private:
+    void TimerThread()
+    {
+        // Pause until timer expired or cancelled
+        while (true)

```

Comment:
is it possible to convert this to a do/while?  Looks like you want some code to execute before the exit condition.